src/ptvsd/server/api.py

```diff
@@ -69,54 +72,116 @@ def debug(address, log_dir=None, multiprocess=True):
+            raise log.exception("{0}() failed:", func.__name__, level="info")
 
     return debug
 
 
 @_starts_debugging
 def enable_attach(dont_trace_start_patterns, dont_trace_end_patterns):
-    if hasattr(enable_attach, "called"):
-        raise RuntimeError("enable_attach() can only be called once per process.")
+    # Errors below are logged with level="info", because the caller might be catching
+    # and handling exceptions, and we don't want to spam their stderr unnecessarily.
+
+    import subprocess
+
+    if hasattr(enable_attach, "adapter"):
+        raise AssertionError("enable_attach() can only be called once per process")
 
     server_access_token = compat.force_str(codecs.encode(os.urandom(32), "hex"))
 
-    import subprocess
+    try:
+        endpoints_listener = sockets.create_server("127.0.0.1", 0, timeout=5)
+    except Exception as exc:
+        log.exception("Can't listen for adapter endpoints:")
+        raise RuntimeError("can't listen for adapter endpoints: " + str(exc))
+    endpoints_host, endpoints_port = endpoints_listener.getsockname()
+    log.info(
+        "Waiting for adapter endpoints on {0}:{1}...", endpoints_host, endpoints_port
+    )
 
     adapter_args = [
         sys.executable,
-        _ADAPTER_PATH,
+        os.path.dirname(adapter.__file__),
         "--for-server",
+        str(endpoints_port),
         "--host",
         options.host,
         "--port",
         str(options.port),
         "--server-access-token",
         server_access_token,
     ]
-
     if log.log_dir is not None:
         adapter_args += ["--log-dir", log.log_dir]
+    log.info("enable_attach() spawning adapter: {0!j}", adapter_args)
+
+    # On Windows, detach the adapter from our console, if any, so that it doesn't
+    # receive Ctrl+C from it, and doesn't keep it open once we exit.
+    creationflags = 0
+    if sys.platform == "win32":
+        creationflags |= 0x08000000  # CREATE_NO_WINDOW
+        creationflags |= 0x00000200  # CREATE_NEW_PROCESS_GROUP
+
+    # Adapter will outlive this process, so we shouldn't wait for it. However, we
+    # need to ensure that the Popen instance for it doesn't get garbage-collected
+    # by holding a reference to it in a non-local variable, to avoid triggering
+    # https://bugs.python.org/issue37380.
+    try:
+        enable_attach.adapter = subprocess.Popen(
+            adapter_args, close_fds=True, creationflags=creationflags
+        )
+        if os.name == "posix":
+            # It's going to fork again to daemonize, so we need to wait on it to
+            # clean it up properly.
+            enable_attach.adapter.wait()
+        else:
+            # Suppress misleading warning about child process still being alive when
+            # this process exits (https://bugs.python.org/issue38890).
+            enable_attach.adapter.returncode = 0
+            pydevd.add_dont_terminate_child_pid(enable_attach.adapter.pid)
```

Comment:
Shouldn't we be adding this in `posix` case as well?

Reply:
In the POSIX case, the immediate child exits immediately after forking itself (and is waited on above). The forked daemon has a different PID, but in any case, it's not treated as a child anymore.